iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp

```diff
@@ -61,41 +67,37 @@ class function_ref;
 ///         callback();
 /// @endcode
 template <class ReturnType, class... ArgTypes>
-class function_ref<ReturnType(ArgTypes...)>
+class function_ref<ReturnType(ArgTypes...)> final
 {
-    using SignatureType = ReturnType(ArgTypes...);
-
-    template <typename T1, typename T2>
-    using has_same_decayed_type = typename std::integral_constant<
-        bool,
-        bool(std::is_same<typename std::decay<T1>::type, typename std::decay<T2>::type>::value)>;
-
   public:
     ~function_ref() noexcept = default;
 
     function_ref(const function_ref&) noexcept = default;
 
-    function_ref& operator=(const function_ref&) noexcept = default;
+    function_ref& operator=(const function_ref&) & noexcept = default;
 
     /// @brief Creates a function_ref with a callable whose lifetime has to be longer than function_ref
     /// @param[in] callable that is not a function_ref
     template <typename CallableType,
-              typename = std::enable_if_t<!is_function_pointer<CallableType>::value
-                                          && !has_same_decayed_type<CallableType, function_ref>::value
-                                          && is_invocable<CallableType, ArgTypes...>::value>>
-    function_ref(CallableType&& callable) noexcept;
+              typename = std::enable_if_t<(!is_function_pointer<CallableType>::value)
+                                          && (!has_same_decayed_type<CallableType, function_ref>::value)
+                                          && (is_invocable<CallableType, ArgTypes...>::value)>>
+    // AXIVION Next Line AutosarC++19_03-A12.1.4 : Implicit conversion is needed for lambdas
```

Comment:
I meant that more accurately it is needed to pass any kind of functors conveniently to functions with `function_ref` arguments. It is not impossible to use lambdas or functors with an explicit version. If it is explicit it just would be inconvenient as the constructor of `function_ref` would have to be explicitly called.

`required to pass lambdas to interfaces with function_ref arguments` would be more accurate.

By the way that is a similar reason as to why the relative pointers work as they do, to avoid inconvenience while passing raw pointers to interfaces or variables that take relative pointers (or generally fancy pointers). This is basically the posterchild example why there need to be exceptions to this explicit ctor rule.